src/dimension.c

```diff
@@ -1646,14 +1694,154 @@ ts_dimension_add(PG_FUNCTION_ARGS)
+
+	if (PG_ARGISNULL(0))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("hypertable cannot be NULL")));
+
+	return ts_dimension_add_internal(fcinfo, &info, false);
+}
+
+TSDLLEXPORT Datum
+ts_dimension_info_in(PG_FUNCTION_ARGS)
+{
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("cannot construct dimension_info from string"),
+			 errdetail("Type dimension_info cannot be constructed from string. You need to "
+					   "use constructor function."),
+			 errhint("Use ByRange or ByHash to construct dimension types.")));
+	PG_RETURN_VOID(); /* keep compiler quiet */
+}
+
+TSDLLEXPORT Datum
+ts_dimension_info_out(PG_FUNCTION_ARGS)
+{
+	DimensionInfo *info = (DimensionInfo *) PG_GETARG_POINTER(0);
+	StringInfoData str;
+	const char *partfuncname =
+		OidIsValid(info->partitioning_func) ? get_func_name(info->partitioning_func) : "-";
+	initStringInfo(&str);
+	switch (info->type)
+	{
+		case DIMENSION_TYPE_CLOSED:
+			appendStringInfo(&str,
+							 "hash//%s//%d//%s",
+							 NameStr(info->colname),
+							 info->num_slices,
+							 partfuncname);
+			break;
+
+		case DIMENSION_TYPE_OPEN:
+		{
+			const char *argvalstr = "-";
+
+			if (OidIsValid(info->interval_type))
+			{
+				bool isvarlena;
+				Oid outfuncid;
+				getTypeOutputInfo(info->interval_type, &outfuncid, &isvarlena);
+				Assert(OidIsValid(outfuncid));
+				argvalstr = OidOutputFunctionCall(outfuncid, info->interval_datum);
+			}
+
+			appendStringInfo(&str,
+							 "range//%s//%s//%s",
+							 NameStr(info->colname),
+							 argvalstr,
+							 partfuncname);
+			break;
+		}
+
+		case DIMENSION_TYPE_ANY:
+			appendStringInfo(&str, "any");
+			break;
+	}
+	PG_RETURN_CSTRING(str.data);
+}
+
+static DimensionInfo *
+make_dimension_info(Name colname, DimensionType dimtype)
+{
+	DimensionInfo *info = palloc0(sizeof(DimensionInfo));
+	info->type = dimtype;
+	namestrcpy(&info->colname, NameStr(*colname));
+	return info;
+}
+
+/*
+ * DimensionInfo for a hash dimension.
+ *
+ * This structure is only partially filled in when constructed. The rest will
+ * be filled in by ts_dimension_add_general.
+ */
+Datum
+ts_hash_dimension(PG_FUNCTION_ARGS)
+{
+	DimensionInfo *info = make_dimension_info(PG_GETARG_NAME(0), DIMENSION_TYPE_CLOSED);
+	Ensure(PG_NARGS() > 2, "expected at most 3 arguments, defined with %d", PG_NARGS());
+	info->num_slices = PG_ARGISNULL(1) ? DatumGetInt32(-1) : PG_GETARG_INT32(1);
+	info->num_slices_is_set = !PG_ARGISNULL(1);
+	info->partitioning_func = PG_ARGISNULL(2) ? InvalidOid : PG_GETARG_OID(2);
+	PG_RETURN_POINTER(info);
+}
+
+/*
+ * DimensionInfo for a hash dimension.
+ *
+ * This structure is only partially filled in when constructed. The rest will
+ * be filled in by ts_dimension_add_general.
+ */
+Datum
+ts_range_dimension(PG_FUNCTION_ARGS)
+{
+	DimensionInfo *info = make_dimension_info(PG_GETARG_NAME(0), DIMENSION_TYPE_OPEN);
+	Ensure(PG_NARGS() > 2, "expected at most 3 arguments, defined with %d", PG_NARGS());
```

Comment:
Do we want to keep these `defined with %d` error messages or should the text be replaced by something like `invoked with %d arguments`?

Reply:
Done.